utils/string/README.md

```diff
@@ -0,0 +1,33 @@
+`icu_string` is a utility crate of the [`ICU4X`] project.
```

Comment:
In Rust, if you want to perform string processing logic, you should almost always use `&str`.  The only time where you care about `String` and `Cow<str>` is if you want to take ownership of the string.  It's not clear to me that this trait is useful in allowing polymorphism over the string storage mechanism; it simply provides some utilities that should already be available if you just use `&str`.

Reply:
Hah thank you for stepping into this rabbit hole with me! That's a reasonable entry point.

The reason I found this trait useful is that our entry point is not str, but Cow. And our output is also a Cow that is a substring of that input.

How would you suggest we write a parser that can take str or Cow which can be owned or borrowed, and produce a Pattern that has Pattern Items that are Cows that can be either owned or borrowed in a more efficient way than with such a trait?

Reply:
> How would you suggest we write a parser that can take str or Cow which can be owned or borrowed, and produce a Pattern that has Pattern Items that are Cows that can be either owned or borrowed in a more efficient way than with such a trait?

Make the pattern parser take an `&'s str` return something that has a lifetime parameter of `'s`.

If the pattern parser *must* take ownership, then it would be more efficient if it would copy the string into a single Cow field and then have Pin references into it.

Splitting a Cow<str> into a bunch of little Cow slices is horribly inefficient if owned, and not as efficient as string slices if borrowed due to the Cow overhead.

Reply:
gotcha, let me see where it takes me. I started splitting the patterns util which uses it since it doesn't overlap with this API, and we can test that on an actual use case there!

Reply:
Yes, this is what I'd recommended earlier: accept `&str`, and if you want you can produce different types, you can have the return type be generic